### DISCUSSION TITLE
Typing plugin helpers

### DISCUSSION THREAD

**Discussion by mfrascati** [nuxt/nuxt#20601] [Questions] [answered]
Hello everybody!
If I understand correctly from the docs, [helpers returned from plugins should be typed automatically](https://nuxt.com/docs/guide/directory-structure/plugins#typing-plugins)

But when I define my plugin

```
import mitt from "mitt";
const emitter = mitt();
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.provide("bus", {
    $on: emitter.on,
    $off: emitter.off,
    $emit: emitter.emit,
  });
});
```

I get $bus is of unknown type (the code works correctly)

<img width="426" alt="image" src="https://user-images.githubusercontent.com/2734747/235427545-d048eac7-9c01-47a0-8277-98af5c097acf.png">

Am I doing something wrong or have I missed something?

Thanks!

**Comment by TheAlexLichter** [comment, marked as answer]
With "returning helpers", the `provide`/`inject` way is meant.

Example from the linked docs:

```ts
export default defineNuxtPlugin(() => {
  return {
    provide: {
      hello: (msg: string) => `Hello ${msg}!`
    }
  }
})
```

Try sth. like the code below or [type it manually](https://nuxt.com/docs/guide/directory-structure/plugins#advanced):

```ts
import mitt from "mitt";

const emitter = mitt();

export default defineNuxtPlugin(() => {
  return {
    provide: {
      bus: {
        $on: emitter.on,
        $off: emitter.off,
        $emit: emitter.emit,
      }
    }
  }
});
```

<img width="997" alt="image" src="https://user-images.githubusercontent.com/640208/235429818-cc68ff64-3607-4105-af53-9fa799395a3a.png">

**Reply by mfrascati** [reply]
Oh sure, I misunderstood the "return" part! Thank you for the quick answer :)

**Reply by TheAlexLichter** [reply]
No problem at all 👏🏻
Glad it works now ☺️
